Add SabreLayout to the C API

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit adds a standalone transpiler pass function to the C API for
running the SabreLayout transpiler pass. This function returns a
new result type QkSabreLayoutResult that contains the output circuit
along with the layouts from running the pass.

### Details and comments

Fixes #14449

TODO:
- [x] Expand test coverage

~This PR is based on top of #14668 and #14715 and will need to be rebased once those two PRs merge. In the meantime you can review just the contents of this PR by looking at the HEAD commit on the PR: https://github.com/Qiskit/qiskit/pull/14711/commits/d1aff2933a4541fecfcc5ee7f429e6ef6c94e00a and https://github.com/Qiskit/qiskit/pull/14711/commits/d08896073fa0615a2fedb7dc491a794cf06bccb9~ This has been rebased now.

mtreinish committed:
Add SabreLayout to the C API

This commit adds a standalone transpiler pass function to the C API for
running the SabreLayout transpiler pass. This function returns a
new result type QkSabreLayoutResult that contains the output circuit
along with the layouts from running the pass.

Fixes #14449

mtreinish commented:
I have a test locally that validates inserting swaps, but it hit an issue that was introduced in #14317. I pushed up a fix in: #14715 I'll embed that into this PR but it will add a dependency on #14715 so this PR will depend on both that and #14668.

mtreinish committed:
Improve testing

mtreinish commented:
I've rebased this PR on main, but it's still depending on: https://github.com/Qiskit/qiskit/pull/14715 which is in the branch. I will rebase it again after that PR merges which is needed to enable the tests as they're written here. I could adjust the tests to not rely on global gates, but I figured since it uncovered a bug that I should leave it using a global 1q gate.

mtreinish commented:
This has been rebased and is unblocked now

mtreinish committed:
Merge remote-tracking branch 'origin/main' into sabre_at_c

mtreinish committed:
Fix tests

mtreinish committed:
Merge branch 'main' into sabre_at_c

mtreinish committed:
Fix docs

mtreinish committed:
Finish API docs for standalone function

mtreinish commented on crates/cext/src/transpiler/passes/sabre_layout.rs:
In updating the docs just now I think we should change the signature here to be:

```suggestion
pub unsafe extern "C" fn qk_transpiler_pass_standalone_sabre_layout(
    circuit: *mut CircuitData,
    target: *const Target,
    max_iterations: usize,
    num_swap_trials: usize,
    num_random_trials: usize,
    seed: i64,
) -> *mut TranspileLayout {
```

Since `SabreLayoutResult` is mostly redundant with `TranspileLayout` and we're already standardizing on it in other places. I will mark this PR as on hold and wait for #14780 to merge and then quickly update this.

jakelishman commented on crates/cext/src/transpiler/passes/sabre_layout.rs:
```diff
@@ -0,0 +1,275 @@
+) -> *mut CircuitData {
+    let result = unsafe { mut_ptr_as_ref(result) };
+    (&mut result.circuit) as *mut _
+}
```
It's not going to be relevant because of the change in direction of the PR, but were it to stay like this, I would have liked to think more about the ownership / lifetime model being implied by this - it's not really clear to me who'd be responsible for freeing the circuit, and how the freeing of the total layout object would have worked under the assumption that the C caller wanted to "move" ownership of the circuit out of this return object.

jakelishman commented on crates/cext/src/transpiler/passes/sabre_layout.rs:
```diff
@@ -0,0 +1,275 @@
+) -> u32 {
+    let result = unsafe { const_ptr_as_ref(result) };
+    result.initial_layout.num_qubits() as u32
+}
```
Again, this will be invalidated by the pivot in return type, but I think this would have always returned the same as `qk_sabre_layout_result_final_layout_num_qubits`, so we probably could have just had one?

jakelishman commented on crates/cext/src/transpiler/passes/sabre_layout.rs:
In the future, we're almost certainly going to want to expose some sort of per-function and per-process "threading configuration" objects into C space, given the target audience.

mtreinish replied:
Yeah, I am putting these comments in the docstrings because I've already had discussions about calling some of the C API functions in an OpenMP context and wanted to make sure we were at least documenting how it worked now. But I agree we'll probably want to expose a structure maybe as part of `TranspileOptions` to control the threading behavior.

jakelishman commented on crates/cext/src/transpiler/passes/sabre_layout.rs:
lol at `QkElidePermutationsResult`

We don't actually really need to _say_ the type here, do we?  It's not a special field in Doxygen, and in C it's necessarily going to have a specified type in the signature to produce a link.

mtreinish replied:
No we don't it's just an old habit from sphinx docs, I've done this in a bunch of the other PRs.

jakelishman commented on crates/cext/src/transpiler/passes/sabre_layout.rs:
This is really neither here nor there, but how come these lines are wrapped at 120 and not 100?

mtreinish replied:
Probably the size of my terminal emulator when I wrote it and then I was relying on `cargo fmt` to line wrap which it didn't do.

jakelishman commented on crates/cext/src/transpiler/passes/sabre_layout.rs:
I don't remember if it's already merged or just in my Sabre PR chain, but I'm fairly sure I change the default at some point.  It doesn't matter a huge amount for right now.

More pressingly: do we want to expose the custom heuristic specification to C now, or are we content to leave it til later?

mtreinish replied:
I was going to leave it until later, mostly because I think there might be a question about integrating it into the transpile function too. We also don't expose it to Python yet do we? Maybe we should do it at the same time.

jakelishman replied:
It's technically exposed to Python in non-stable API (*edit*: in SabreSwap, at least). But yeah, I'm fine leaving it to a later interface, since we're not committing to stability here.

jakelishman commented on crates/cext/src/transpiler/passes/sabre_layout.rs:
This `match` seems like a slightly odd way of writing `sabre_layout_and_routing(...).unwrap()`?

mtreinish replied:
I switched it to unwrap in: https://github.com/Qiskit/qiskit/pull/14711/commits/0aa1f10b3d2d4c84e5f01d9c4eb7d5991e8ffd69

jakelishman reviewed (commented):
I'll look again after the interface change to `QkTranspileLayout`.  There's a few places in this PR (and maybe others?) where we're doing `match result { Ok(x) => x, Err(e) => panic!("{e}") }`, which I think is just `Result::unwrap`, but with `Display` formatting instead of `Debug`?

mtreinish commented:
> I'll look again after the interface change to `QkTranspileLayout`. There's a few places in this PR (and maybe others?) where we're doing `match result { Ok(x) => x, Err(e) => panic!("{e}") }`, which I think is just `Result::unwrap`, but with `Display` formatting instead of `Debug`?

Yeah that's fair, I was thinking there was a header in the message with something about `unwrap()` but I was misremembering or confusing it with something else. Regardless I probably need to change this to have an explicit error string since we'll get a panic inside the panic handling if it's a `PyErr`.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into sabre_at_c

mtreinish committed:
Pivot interface to use TranspileLayout

This commit reworks the C interface for sabre layout to use
TranspileLayout for the return type remove the Sabre result type. It was
largely duplicative and it's better to just use the same interface.

mtreinish committed:
Remove stale references to QkSabreLayoutResult in docs

jakelishman commented on crates/cext/src/transpiler/passes/sabre_layout.rs:
We don't need to do it for this PR, but for the future when we're moving to a more stable interface, we might want to factor some of these things out (and the heuristic) into a `SabreLayoutConfig` object or something, with a C-space builder interface, so it's more resistant to API changes.  (I tried out something like this with VF2 in https://github.com/Qiskit/qiskit/pull/14861/commits/5779fd92130946f582c38641e1169338d70104df).

mtreinish replied:
I am going to switch to this, I do like the options struct pattern a lot this is a good idea and I think it makes sense here too, especially if we want to expose heuristic in the future. It'll make the api change easier, especially if we add a default function for building them.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14711/commits/dcc672b1767d67aec1eb76312b5eab35bdb0594a

jakelishman commented on crates/cext/src/transpiler/passes/sabre_layout.rs:
In C land, it might be cleaner to enforce that a seed _must_ be set, and take this as `u64` to allow access to the native size?

mtreinish replied:
I was assuming there were people who wanted to seed form system entropy. I guess you can do that fairly easily in C (although you can in Python too). I'm not opposed to requiring a seed necessarily I would assume most people want one for reproducibility.

mtreinish replied:
With the above comment I made seed a u64 in the options struct, and just had the default function for building the options, seed the seed from system entropy (I need to work seed into that sentence 3 more times to win a prize).

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14711/commits/dcc672b1767d67aec1eb76312b5eab35bdb0594a

jakelishman commented on crates/cext/src/transpiler/passes/sabre_layout.rs:
Stylistic: I think `expect` messages are typically a reason _why_ the author expected this to work without panicking, rather than a failure message that's typically just going to reproduce the stack trace.

mtreinish replied:
I can change it to a match and panic, that's what it was before, This just seemed the most concise way to convert the `PyErr` to a user facing error message.

For the most part these should be pretty hard to trigger form the api exposed in C.

jakelishman replied:
I'm fine with `expect`, I was more just meaning that stylistically, and especially if it's not intended to trigger, `expect` messages are more like an explanation for _why_ you don't think it'll trigger.  Btw, `Result::expect` still requires the error type to implement `Debug` - is that a problem for `PyErr` too?

jakelishman commented on crates/cext/src/transpiler/passes/sabre_layout.rs:
Returning a `TranspileLayout` like this discounts the possibility that the user might already have an "output permutation" they care about tracking (c.f. `Split2qUnitaries` or `ElidePermutations`).  I'm fine if we decide for now that "standalone means it stands _entirely_ alone" and so there's no other context (before `TranspileLayout` moves into `CircuitData`/`DAGCircuit` properly), I'm mostly just flagging this decision for the record.

mtreinish replied:
We didn't expose the API for permutation composition to C. We could add it I guess, but I think for right now the intent is for the standalone functions to be standalone. That's the position we took for the other C standalone pass functions too. We'll need to think about the larger workflows for building custom pass managers in >=2.3.

jakelishman commented on crates/cext/src/transpiler/passes/sabre_layout.rs:
The last argument to `TranspileLayout` is the number of input qubits, but at this point we've just overridden `circuit` with `out_circuit`, which is ancilla-expanded, so I think we're feeding it the wrong number.

mtreinish replied:
Oops, good catch I wrote the `*circuit = out_circuit` line after this  (I realized I forgot it) and missed this. I'll fix this.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14711/commits/be8e750e6308e370ff8847ebcc05369ad4fdd272

jakelishman commented on crates/cext/src/transpiler/passes/sabre_layout.rs:
I'm fairly sure the `final_layout` that comes out of Sabre in Rust space is not the same sort of quasi-permutation that used to come out of Sabre in Python space.  I think what we need here is something like `(0..num_qubits).map(|q| Qubit::from(final_layout[initial_layout[PhysicalQubit(q)]])).collect()` or something like that. (Or to add a separate constructor to `TranspileLayout` that handles it for us.)

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14711/commits/cb8960d8d91968111c8969261cb93f26a03147d3

jakelishman commented on docs/cdoc/index.rst:
```diff
@@ -68,3 +68,4 @@ results transpiling circuits from Python via the C API.
    qk-transpiler-passes
    qk-vf2-layout-result
    qk-elide-permutations-result
+   qk-sabre-layout-result
```
Revert?

mtreinish replied:
Removed in: https://github.com/Qiskit/qiskit/pull/14711/commits/05725baf319e99b410e53b1af021d38d430dc3ec

jakelishman commented on docs/cdoc/qk-sabre-layout-result.rst:
```diff
@@ -0,0 +1,18 @@
+===================
+QkSabreLayoutResult
+===================
```
Can this file be reverted?

mtreinish replied:
Removed in: https://github.com/Qiskit/qiskit/pull/14711/commits/05725baf319e99b410e53b1af021d38d430dc3ec

jakelishman commented on test/c/test_sabre_layout.c:
```diff
@@ -0,0 +1,325 @@
+    }
+}
+
+bool compare_circuits(const QkCircuit *res, const QkCircuit *expected) {
```
These guys _maybe_ want to live in a test-helpers module.  `compare_circuits` probably ought to be promoted to proper C API and use the `__eq__` method internally / a new `CircuitData::structurally_equal` method?  We can follow up on that, if we make a decision.

mtreinish replied:
I moved them into a common location in: https://github.com/Qiskit/qiskit/pull/14711/commits/6bf9a6e53a3a2ee327b3a86d1148cd2294488e84

I think using `__eq__` is what we'll want longer term but there will need to be some work on decoupling it from Python first.

jakelishman commented on test/c/test_sabre_layout.c:
```diff
@@ -0,0 +1,325 @@
+        return false;
+    }
+    QkCircuitInstruction *res_inst = malloc(sizeof(QkCircuitInstruction));
+    QkCircuitInstruction *expected_inst = malloc(sizeof(QkCircuitInstruction));
```
why `malloc` and not stack vars?

mtreinish replied:
I assume I copied this from somewhere. I think I had circuit tests I wrote like this mostly to test `sizeof()`. I'll change it to use stack vars instead.

mtreinish replied:
Changed in: https://github.com/Qiskit/qiskit/pull/14711/commits/87e98f195728f976554059b8953db516e03a8b92

jakelishman commented on test/c/test_sabre_layout.c:
```diff
@@ -0,0 +1,325 @@
+        int result = strcmp(res_inst->name, expected_inst->name);
+        if (result != 0) {
+            printf("Gate %d have different gates %s was found and expected %s", i, res_inst->name,
+                   expected_inst->name);
```
It's a test so this probably doesn't matter at all, but should we be `fprintf(stderr, ...)`?

Also, the format string probably wants a `\n` at the end and the grammar is a bit weird.

jakelishman commented on test/c/test_sabre_layout.c:
```diff
@@ -0,0 +1,325 @@
+    for (size_t i = 0; i < qk_circuit_num_instructions(res); i++) {
+        qk_circuit_get_instruction(res, i, res_inst);
+        qk_circuit_get_instruction(expected, i, expected_inst);
+        int result = strcmp(res_inst->name, expected_inst->name);
```
Raw `strcmp` on two pointers where neither of them are string literals makes me vaguely nervous (what if we accidentally return nonsense with no null byte from C?), but equally, this is a test we entirely control so the risk is not high.

(You can ignore this: I'm mostly just flapping around a bit trying to develop more intuition for good C practices.)

jakelishman commented on test/c/test_sabre_layout.c:
```diff
@@ -0,0 +1,325 @@
+                free(expected_inst);
+                return false;
+            }
+        }
```
It's not going to matter in this test suite, but this is asking for floating-point errors.

jakelishman commented on test/c/test_sabre_layout.c:
```diff
@@ -0,0 +1,325 @@
+            result = EqualityError;
+            goto layout_cleanup;
+        }
+    }
```
Oh good, another place to update when Sabre's randomisation changes lol

mtreinish replied:
lol, I was thinking the same thing when I wrote these tests. But I couldn't think of a better way to test this that actually uses the `TranspileLayout`.

jakelishman replied:
Yeah, I don't have a better solution either

jakelishman commented on test/c/test_sabre_layout.c:
```diff
@@ -0,0 +1,325 @@
+        }
+    }
+    QkCircuit *expected_circuit = qk_circuit_new(5, 0);
+    uint32_t qargs[2] = {1, 0};
+    qk_circuit_gate(expected_circuit, QkGate_CX, qargs, NULL);
+    qargs[0] = 1;
+    qargs[1] = 2;
+    qk_circuit_gate(expected_circuit, QkGate_CX, qargs, NULL);
```
Arrays being non-assignable kind of sucks.  I wonder if
```c
uint32_t qargs[2] = {1, 0};
qk_circuit_gate(expected, QkGate_CX, qargs, NULL);
uint32_t qargs[2] = {1, 2};
qk_circuit_gate(expected, QkGate_CX, qargs, NULL);
// ...
```
would be a nicer pattern when it comes to writing docs?  

Though now I mention it, perhaps the fact that we pass a pointer into the stack to `qk_circuit_gate` will cause a C compiler to think that the memory for each `qargs` needs to stay alive until the end of the function, and we'll end up with a humongous stack frame?

Either way, I think we need to work out a neater way to write this sort of thing for documentation in the future, or potentially to provide specialised 2q standard-gate methods, or _something_ to ease the API.

mtreinish replied:
Heh, every time I context switch to C I do the same thing, first try reassigning the array like:
```c
qargs = {1, 2}
```
then I try the redefinition like you're suggesting. But it's a compiler error for me locally with "redefinition of ‘qargs’". Then I write it like this. I'm not sure of a better syntax here, but my C is fairly rusty.

We can look at expanding the specialized methods, I think that's really the only other option here really.

jakelishman commented on test/c/test_sabre_layout.c:
```diff
@@ -0,0 +1,325 @@
+    if (!compare_result) {
+        result = EqualityError;
+        goto layout_cleanup;
+    }
```
Mostly unimportant, but is this the standard C return-value convention for a comparator?  Things like `strcmp` and process return codes return 0 on match/success, but we're returning 0 (`false`) for the failure case.

mtreinish replied:
I can invert the logic, I didn't really think about it too much when I wrote the helper function.

jakelishman commented on crates/cext/src/transpiler/passes/sabre_layout.rs:
Do we have any C-level objects that borrow a reference from a pointer?  If so, we probably should get in the habit of documenting in passes like this one that calling frees the underlying `CircuitData` object, and so can only be called with the last extant reference to a `CircuitData`.

mtreinish replied:
Yeah that's a fair point. This is the pattern most of the standalone transpiler passes use for `CircuitData` so it's worth documenting this so users no to copy the circuit if they need it to persist past calling the pass function.

mtreinish replied:
Added in: https://github.com/Qiskit/qiskit/pull/14711/commits/6209b8d2ebdccea02806741a2cb0fed6cecd5e18

jakelishman commented on test/c/test_sabre_layout.c:
```diff
@@ -0,0 +1,325 @@
+    QkTranspileLayout *layout_result =
+        qk_transpiler_pass_standalone_sabre_layout(qc, target, 4, 20, 20, 2025);
+
+    QkOpCounts op_counts = qk_circuit_count_ops(qc);
```
I don't think we ever free this.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14711/commits/6d95b87cbf14fba5993c6dc4390ca68dffd1f83e

jakelishman commented on test/c/test_sabre_layout.c:
```diff
@@ -0,0 +1,325 @@
+            goto layout_cleanup;
+        }
+    }
+    QkCircuit *expected_circuit = qk_circuit_new(5, 0);
```
I don't think we ever free this.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14711/commits/6d95b87cbf14fba5993c6dc4390ca68dffd1f83e

jakelishman commented on test/c/test_sabre_layout.c:
```diff
@@ -0,0 +1,325 @@
+    }
+    QkTranspileLayout *layout_result =
+        qk_transpiler_pass_standalone_sabre_layout(qc, target, 4, 20, 20, 42);
+    QkCircuit *expected_circuit = qk_circuit_new(5, 0);
```
Memory leak on `expected_circuit`?

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14711/commits/6d95b87cbf14fba5993c6dc4390ca68dffd1f83e

mtreinish committed:
Fix test memory leaks

mtreinish committed:
Make input to pass function an options struct

mtreinish committed:
Set number of input qubits correctly in output layout

mtreinish committed:
Move helper functions to common.c

mtreinish committed:
Use a stack variable instructions in compare_circuits()

mtreinish committed:
Fix output_permutation in the return

mtreinish committed:
Merge branch 'main' into sabre_at_c

mtreinish committed:
Merge branch 'main' into sabre_at_c

jakelishman commented on crates/cext/src/transpiler/passes/sabre_layout.rs:
```diff
@@ -0,0 +1,161 @@
+///
+/// Build a default sabre layout options object This builds a sabre layout with ``max_iterations``
+/// set to 4, both ``num_swap_trials`` and ``num_random_trials`` set to 20, and the seed selected
+/// by a RNG seeded from system entropy.
```
```suggestion
/// Build a default sabre layout options object. This builds a sabre layout with ``max_iterations``
/// set to 4, both ``num_swap_trials`` and ``num_random_trials`` set to 20, and the seed selected
/// by a RNG seeded from system entropy.
```

jakelishman commented on crates/cext/src/transpiler/passes/sabre_layout.rs:
```diff
@@ -0,0 +1,161 @@
+use qiskit_transpiler::transpile_layout::TranspileLayout;
+
+/// The options for running ``qk_transpiler_pass_standalone_sabre_layout``. This struct is used
+/// as an input to control the behavior of the
```
```suggestion
/// The options for running ``qk_transpiler_pass_standalone_sabre_layout``. This struct is used
/// as an input to control the behavior of the layout and routing algorithms.
```

?

jakelishman commented on crates/cext/src/transpiler/passes/sabre_layout.rs:
```diff
@@ -0,0 +1,161 @@
+
+/// The options for running ``qk_transpiler_pass_standalone_sabre_layout``. This struct is used
+/// as an input to control the behavior of the
+#[repr(C)]
+pub struct QkSabreLayoutOptions {
+    /// The number of forward-backward iterations in the sabre routing algorithm
+    max_iterations: usize,
+    /// The number of trials to run of the sabre routing algorithm for each iteration. When > 1 the
+    /// trial that routing trial that results in the output with the fewest swap gates will be
+    /// selected.
+    num_swap_trials: usize,
+    /// The number of random layout trials to run. The trial that results in the output with the
+    /// fewest swap gates will be selected.
+    num_random_trials: usize,
+    /// A seed value for the pRNG used internally.
+    seed: u64,
+}
```
In VF2 land I went for making the internals of the struct private to let us extend it in an API-safe way, but I think it's good to try both ways and see which feels better, while we're still uncommitted to a stable interface - let's leave this as-is.

jakelishman commented on crates/cext/src/transpiler/passes/sabre_layout.rs:
Did we decide that `<PyErr as Debug>` is a problem too?  So `PyResult::expect` is unusable as well.

mtreinish replied:
Yeah, both debug and display need the gil: https://github.com/PyO3/pyo3/blob/v0.25.1/src/err/mod.rs#L658-L696 so back to the match and panic I guess.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14711/commits/59d316d9e095855c7d0f66e5cbe5e7bab1cc491d

jakelishman commented on crates/cext/src/transpiler/passes/sabre_layout.rs:
The `unwrap` and the `expect` both use `PyErr as Debug`.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14711/commits/59d316d9e095855c7d0f66e5cbe5e7bab1cc491d

jakelishman commented on test/c/common.c:
```diff
@@ -36,3 +38,102 @@ int run(const char *name, int (*test_function)(void)) {
+
+bool compare_circuits(const QkCircuit *res, const QkCircuit *expected) {
+    if (qk_circuit_num_instructions(res) != qk_circuit_num_instructions(expected)) {
+        printf("Number of instructions in circuit is mismatched");
```
(Almost?) all of the `printf` in this function are missing the trailing newline.

mtreinish replied:
You know I don't think I'm ever going to remember this Python and rust integrating the newline has spoiled me.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14711/commits/4ef73e455a7b1d73e58b470228fa1789be343bcf

jakelishman commented on test/c/test_sabre_layout.c:
```diff
@@ -0,0 +1,216 @@
+            result = EqualityError;
+            goto layout_cleanup;
+        }
+    }
```
The text and variables here are about `final_layout`, but the assertions are about the output permutation.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14711/commits/fd4e06d21fcddf93c63e8d01a29b773680ac857d (well the text, I left the variables because I missed that part of your comment)

jakelishman commented on test/c/test_sabre_layout.c:
```diff
@@ -0,0 +1,216 @@
+            result = EqualityError;
+            goto layout_cleanup;
+        }
+    }
```
Same here, with the text being about layout but the assertions being about permutation.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14711/commits/fd4e06d21fcddf93c63e8d01a29b773680ac857d

jakelishman commented on test/c/test_sabre_layout.c:
Also most of the `printf` in the file are missing newlines.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14711/commits/4ef73e455a7b1d73e58b470228fa1789be343bcf

jakelishman reviewed (commented):
Minor stuff

mtreinish committed:
Apply suggestions from code review

Co-authored-by: Jake Lishman <jake@binhbar.com>

mtreinish committed:
Update docstring for circuit arg

mtreinish committed:
Don't use PyErr debug or display

mtreinish committed:
Add missing newlines to test failure strings

mtreinish committed:
Update error messages about output permutation test failures

jakelishman committed:
Correct permutation variable names

jakelishman reviewed (dismissed):
Thanks!

jakelishman committed:
Correct copyright header
